pkg/objstore/s3/s3.go

```diff
@@ -90,12 +90,13 @@ func NewBucketWithConfig(logger log.Logger, config Config, component string) (*B
 	if err := validate(config); err != nil {
 		return nil, err
 	}
-	if config.AccessKey != "" {
-		signature := credentials.SignatureV4
-		if config.SignatureV2 {
-			signature = credentials.SignatureV2
-		}
 
+	signature := credentials.SignatureV4
```

Comment:
Maybe in theory, but unfortunately not in practice. Using `v0.2.1` with the following `bucket.yaml` and environment variables (`AWS_ACCESS_KEY_ID` and `AWS_SECRET_ACCESS_KEY`), in AWS:

bucket.yaml
```
type: S3
config:
  bucket: my-secret-bucket-name
  endpoint: s3.us-west-2.amazonaws.com
  insecure: false
  signature_version2: false
  encrypt_sse: false
```
 fails with the following error:
```
level=info ts=2019-01-15T19:38:03.422651711Z caller=flags.go:75 msg="gossip is disabled"
level=info ts=2019-01-15T19:38:03.422720714Z caller=factory.go:39 msg="loading bucket configuration"
store command failed: bucket store initial sync: sync block: iter: Access Denied
```

With this patch it works correctly.

Reply:
Hm.. I think I know what is happening (: 

So @jojohappy is right here, nothing really changed by moving signature outside of `if config.AccessKey` body.

I think this patch forks for you because recently we fixed S3 auth order: https://github.com/improbable-eng/thanos/pull/732 This was after 0.2.1 was released (: I think you accidenly assumed your patch work, but `master` most likely works for you as well (: